### PR TITLE
✨ Cairo 1 Validation

### PR DESCRIPTION
**Pull request by EvolveArt**
Solves #7 #18 #19

This PR adds full support for `v2.1.0` config.
It also adds multi-threaded validation of transactions which drastically speeds up the process. Note that we limit to 95 tasks because Madara prevents more than 100 simultaneous connections.
It also moves transactions' validation after building reports to not skew results.

It also fixes few bugs and cleans some code.

Next step will be adding multi-threading to send transactions. Weirdly, Cairo 1 accounts are like 10x slower than Cairo 0 accounts to send transactions which I haven't understood why yet. It could come from the execution encoding in `starknet-rs` which is the only thing that really changes apart from the class hash.

### CODE REVIEW

File: src/utils.rs [hunk 1]
```diff
@@ -143,23 +142,21 @@ pub async fn wait_for_tx(
             }
             Ok(PendingReceipt(_)) => {
                 debug!("Waiting for transaction {tx_hash:#064x} to be accepted");
-                tokio::time::sleep(WAIT_FOR_TX_SLEEP).await;
+                tokio::time::sleep(check_interval).await;
```
[the remark this answers is not in the document]
Comment (EvolveArt) [reply src/utils.rs:145]: I prefer to have it as a param

File: src/actions/shoot.rs [hunk 2]
```diff
@@ -62,16 +65,16 @@ pub struct GatlingShooter {
...
-    account: SingleOwnerAccount<Arc<JsonRpcClient<HttpTransport>>, LocalWallet>,
+    account: StarknetAccount,
     nonces: HashMap<FieldElement, FieldElement>,
     environment: Option<GatlingEnvironment>, // Will be populated in setup phase
 }
 
 #[derive(Clone)]
 pub struct GatlingEnvironment {
-    erc20_address: FieldElement,
+    _erc20_address: FieldElement,
```
[the remark this answers is not in the document]
Comment (EvolveArt) [reply src/actions/shoot.rs:75]: not being used for now as we would need to fund all the accounts with that ERC20 at deployment. Rather just use the fee token.

File: src/actions/shoot.rs [hunk 3]
```diff
@@ -213,18 +217,43 @@ impl GatlingShooter {
...
-            match wait_for_tx(&self.starknet_rpc, transaction, CHECK_INTERVAL).await {
-                Ok(_) => {
-                    accepted_txs.push(transaction);
-                    debug!("Transaction {:#064x} accepted", transaction)
+        // Verify transactions in parallel
+        let chunk_size = transactions.len() / 10;
+        let transactions_sets = transactions.chunks(chunk_size).map(|chunk| chunk.to_vec());
+
+        let fetches = futures::stream::iter(transactions_sets.map(|transactions_set| {
+            // Should we clone the rpc client or is it ok to share it between threads ?
```
[the remark this answers is not in the document]
Comment (EvolveArt) [reply src/actions/shoot.rs:225]: Alright then we can keep it like this for now

File: src/actions/shoot.rs [hunk 4]
```diff
@@ -213,18 +217,43 @@ impl GatlingShooter {
         let mut accepted_txs = Vec::new();
         let mut errors = Vec::new();
 
-        for transaction in transactions {
-            match wait_for_tx(&self.starknet_rpc, transaction, CHECK_INTERVAL).await {
-                Ok(_) => {
-                    accepted_txs.push(transaction);
-                    debug!("Transaction {:#064x} accepted", transaction)
+        // Verify transactions in parallel
+        let chunk_size = transactions.len() / 10;
```
[the remark this answers is not in the document]
Comment (EvolveArt) [reply src/actions/shoot.rs:221]: no as we divide the work in 10 chunks such that each connection will be responsible for one chunk. Will make it a constant that will be the same as `buffer_unordered`

[the remark this answers is not in the document]
Comment (d-roak) [reply src/actions/shoot.rs:221]: ack

**Pull request merged**